### ISSUE TITLE
Wired handling of host "localhost" in config.php

### ISSUE THREAD

**Issue by tacruc** [nextcloud/notify_push#16]
I had two issues getting this up and running, both correlated of host's being called localhost and not 127.0.0.1.

1. Mysql connection failed. As the mysql user was only allowed to login to localhost but not 127.0.0.1. So notify_push didn't use the proper host of config.php to login.
2. redis host being localhost led to a `0: Connection refused (os error 111)`  changing it to 127.0.0.1 in the config.php solved the issue.

So they might not be connected, but both errors seem to be connected to different handling of the host value "localhost" of php and rust. I mean the config was working with php over the last years.

**Comment by icewind1991** [comment]
Can you provide the return of the following mysql query: `show variables like 'socket';`

**Comment by tacruc** [comment]
```
Variable_name | Value |  
socket | /run/mysql/mysql.sock
```

But I don't see how this might be connected?
For me the main problem is that the binary and the php differently use translate the config.

**Comment by icewind1991** [comment]
> But I don't see how this might be connected?

php transparently uses the mysql socket when trying to connect to `localhost`, in order for the binary to do the same it needs to know where the sockets might be.

I dont think php-redis does the same, maybe that issue is related to `localhost` resolving to ipv6, can you try running the binary with the `LOG=notify_push=trace` environment variable and posting the dump of the parsed config it provides (obfuscating the password)

Something like

```
LOG=notify_push=trace apps/notify_push/bin/x86_64/notify_push config/config.php
```

when ran from the main nextcloud folder

**Comment by tacruc** [comment]
First of all ping localhost translate to ipv6.
Log is following shortly.

**Comment by tacruc** [comment]
```
[2021-02-24 18:01:48.141394 +01:00] TRACE [notify_push] src/main.rs:41: Running with config: Config { database: AnyConnectOptions(MySql(MySqlConnectOptions { host: "localhost", port: 3306, socket: None, username: "owncloud", password: Some("****"), database: Some("owncloud"), ssl_mode: Preferred, ssl_ca: None, statement_cache_capacity: 100, charset: "utf8mb4", collation: None, log_settings: LogSettings { statements_level: Info, slow_statements_level: Warn, slow_statements_duration: 1s } })), database_prefix: "oc_", redis: ConnectionInfo { addr: Tcp("localhost", 6379), db: 0, username: None, passwd: None }, nextcloud_url: "****" }
[2021-02-24 18:01:48.142910 +01:00] DEBUG [notify_push::storage_mapping] src/storage_mapping.rs:96: querying storage mapping for 1
Error: 
   0: Failed to connect to redis
   1: Connection refused (os error 111)

Location:
   /tmp/krankerl.6wDVzxnHMC1t/notify_push/src/lib.rs:112

Backtrace omitted.
Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
```

**Comment by icewind1991** [comment]
Probably caused by it trying to use ipv6 then, I'll set it to always use ipv4 when `localhost` is specified.

**Comment by tacruc** [comment]
Is it hard to catch the error and try both? Although, it is probably  unlikely that redis is only listening for ipv6.

**Comment by icewind1991** [comment]
I can try that if the current solution becomes a problem, for now this solution adds less complexity.

Fix is released with 0.1.4

**Comment by q-wertz** [comment]
@tacruc could you solve this issue? Having the same problem... 
```
Error: 
  × Failed to connect to database
  ├─▶ error returned from database: 1130: Host '127.0.0.1' is not allowed to connect to this MariaDB server
  ╰─▶ 1130: Host '127.0.0.1' is not allowed to connect to this MariaDB server
```

What helps is changing the recommended config from the [Nextcloud admin manual](https://docs.nextcloud.com/server/26/admin_manual/configuration_database/linux_database_configuration.html) and uncomment
```
skip_name_resolve = 1
```

But not sure if this comes with security issues...
